File: security/redis_lock.go
package security

import (
	"context"
	"fmt"
	"sync"
	"time"
)

type DistributedLockManager struct {
	mu     sync.Mutex
	locks  map[string]time.Time
	ttl    time.Duration
}

func NewDistributedLockManager(ttl time.Duration) *DistributedLockManager {
	if ttl <= 0 {
		ttl = 30 * time.Second
	}
	return &DistributedLockManager{
		locks: make(map[string]time.Time),
		ttl:   ttl,
	}
}

func (m *DistributedLockManager) AcquireLock(ctx context.Context, lockKey string) (bool, error) {
	m.mu.Lock()
	defer m.mu.Unlock()

	expiry, exists := m.locks[lockKey]
	now := time.Now()

	// If lock exists and hasn't expired, acquisition fails
	if exists && now.Before(expiry) {
		return false, nil
	}

	// Acquire lock with TTL
	m.locks[lockKey] = now.Add(m.ttl)
	return true, nil
}

func (m *DistributedLockManager) ReleaseLock(ctx context.Context, lockKey string) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	delete(m.locks, lockKey)
	return nil
}

func (m *DistributedLockManager) IsLocked(ctx context.Context, lockKey string) bool {
	m.mu.Lock()
	defer m.mu.Unlock()
	expiry, exists := m.locks[lockKey]
	return exists && time.Now().Before(expiry)
}

func FormatVerificationLockKey(doctorID string) string {
	return fmt.Sprintf("lock:verification:doctor:%s", doctorID)
}

File: backend_api/security/redis_lock.py
import time
from threading import Lock
from typing import Dict

class DistributedLockManager:
    def __init__(self, ttl: float = 30.0):
        if ttl <= 0:
            ttl = 30.0
        self.mu = Lock()
        self.locks: Dict[str, float] = {}
        self.ttl = ttl

    def acquire_lock(self, context, lock_key: str) -> bool:
        with self.mu:
            expiry = self.locks.get(lock_key)
            now = time.time()

            if expiry is not None and now < expiry:
                return False

            self.locks[lock_key] = now + self.ttl
            return True

    def release_lock(self, context, lock_key: str):
        with self.mu:
            if lock_key in self.locks:
                del self.locks[lock_key]

    def is_locked(self, context, lock_key: str) -> bool:
        with self.mu:
            expiry = self.locks.get(lock_key)
            if expiry is not None and time.time() < expiry:
                return True
            return False

def format_verification_lock_key(doctor_id: str) -> str:
    return f"lock:verification:doctor:{doctor_id}"
